### ISSUE TITLE
nul-terminated array bounds checks

### ISSUE THREAD

**Issue by mwhicks1** [Microsoft/checkedc#235]
I think I'm triggering some bounds checks. Two questions:

1) The bounds check failure is manifesting as SIGILL, "illegal instruction". How should this be manifesting, instead?
```
% ./a.out 1 2 2 33 4 345 34 3 453 43545
Illegal instruction: 4
```
2) It seems that the rules could be relaxed to allow not just reading the last position but also writing a `'\0'` there. Any reason not to allow that? That's what's causing my problem right now.

**Comment by dtarditi** [comment]
I'll address question 2 first.  I don't want my response to get buried  I have more say about 1 than you might expect.

The main cost will be that bounds checks for writes to null-terminated arrays will get a little more expensive in terms of code size.   Given bounds of `bounds(lb, ub)` and a pointer `p`, the bounds checks for writing a value 'v' to a null-terminated array is `lb <= p < ub`   It would be `lb <= p < ub || (v = 0 && p == ub).`     It won't affect code speed much because the common case is still covered by the first clause.  The second clause covers an exceptional case.

I think this will be a pretty common error to make.  I made it in the sample code that I wrote and you made it, and I'm sure lots of other programmers will make it.  We could try to statically detect when a value is being written at the upper bound and warn the programmer that they need to insert a check.  Of course, a static check can still miss cases. 

So the arguments for and against relaxing the rules:
- For: this will be a common error and there won't be a speed effect in the common case.
- Against: there's a potential code size increase.  Programmers can always add explicit checks to avoid writing null at the last location, and we can warn when they seem to have missed a check in common cases.

For 1, this is what's expected right now.   We're using the LLVM trap instruction to signal a runtime fault and that's the behavior.  Eventually I think we should produce a SIGSEGV.    Is the use of SIGILL causing problems with debugging or not getting core dumps?

This issue of how to trap efficiently and give programmers information for debugging is a little more complicated than it seems.   We do introduce lots of trap instructions.   In optimized code they get combined right now, but it isn't clear that you want that behavior in production systems.  When they are combined, you don't know where in a function the buffer overrun happened, which makes debugging challenging.

I think the LLVM folks worked through the issues for traps and decided that an illegal instruction for trap was the one thing that they could do that works across all platforms and is  efficient. It is a 1-byte instruction on x86.  SIGSEGV require a little more code and OS-specific behavior.  You can, for example, store a value to 0, assuming the first page of memory is protected.   If you want to be truly portable, you'll need to insert calls to the raise function, but that starts to get costly in terms of code size.

**Comment by mwhicks1** [comment]
I think you should relax the rules for 2, at least until you've worked out how to initialize NT pointers safely. What is happening now is that I'm doing a `malloc()`, and then trying to write the `'\0'` to make it well-formed. When I do that a dynamic check is failing, which is ironic because if I didn't write the `'\0'` I'd be in even more trouble. It might be that you want this check even when initialization is worked out. If you have a way to work around this problem, I'm happy to know it (for your `string-examples.c` file, you never allocate strings, but rather use ones from `argv` so that didn't help me).

For 1, having `SIGILL` is fine. I just didn't know that this is what should happen. I expected to be told which bounds check failed, and where.

**Comment by dtarditi** [comment]
I agree that we should allow programs to write null terminators at the upper bounds of null-terminated array.   It's safe and will prevent programming errors.   Both you and I made errors when converting existing code where the code we converted tried to write a null exactly at the upper bound.  The code had the pattern of copying N characters and then writing a null terminator.  For that case N + 1 bytes are allocated, given the null-terminated array a count of N.    The null-terminator itself as at the upper bound N.
